### PR TITLE
revamp entire metrics

### LINKED ISSUE #3877: merge new metrics API

### PR DESCRIPTION
**Pull request by ananyahjha93**
Co-authored-by: Teddy Koker <teddy.koker@gmail.com>

## What does this PR do?
Deletes metrics

Fixes #3877
# Before submitting

- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests? 
- [X] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Always 💯  
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/metrics/functional/test_classification.py [hunk 1]
```diff
@@ -1,485 +0,0 @@
-from functools import partial
```
Comment (justusschock) [inline tests/metrics/functional/test_classification.py:1]: If we keep functional metrics, we also should keep functional tests

File: tests/metrics/functional/test_nlp.py [hunk 2]
```diff
@@ -1,66 +0,0 @@
-import pytest
```
Comment (justusschock) [inline tests/metrics/functional/test_nlp.py:1]: keep functional tests if we keep metrics

File: tests/metrics/functional/test_reduction.py [hunk 3]
```diff
@@ -1,30 +0,0 @@
-import pytest
```
Comment (justusschock) [inline tests/metrics/functional/test_reduction.py:1]: keep functional tests if we keep metrics

File: tests/metrics/functional/test_regression.py [hunk 4]
```diff
@@ -1,175 +0,0 @@
-import numpy as np
```
Comment (justusschock) [inline tests/metrics/functional/test_regression.py:1]: keep functional tests if we keep metrics

File: tests/metrics/functional/test_self_supervised.py [hunk 5]
```diff
@@ -1,35 +0,0 @@
-import pytest
```
Comment (justusschock) [inline tests/metrics/functional/test_self_supervised.py:1]: keep functional tests if we keep metrics

**Review by justusschock — COMMENTED** [review]
I really like the new interface. I only had some minor comments (e.g. let's keep functional metrics for now, some naming stuff and some type annotations). I assume docstrings will be added later?

File: pytorch_lightning/metrics/metric.py [hunk 6]
```diff
@@ -1,262 +1,148 @@
...
+
+            # restore context
+            for attr, val in self._cache.items():
+                setattr(self, attr, val)
+            self._to_sync = True
+            self._computed = None
+
+            return result
+
+    def sync(self):
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/metrics/metric.py:70]: @justusschock someone had a PR open for changing this to sync_dist because they were using it with horovod. i like either sync or sync_dist

[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py:70]: Sync_dist is also fine with me. I just prefer something more meaningful than just sync

File: pytorch_lightning/metrics/metric.py [hunk 7]
```diff
@@ -1,262 +1,148 @@
...
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-
+import functools
 from abc import ABC, abstractmethod
-from typing import Any, Mapping, Optional, Sequence
-import numbers
+from typing import Any, Callable, Optional, Union
+from collections.abc import Mapping, Sequence
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/metrics/metric.py:4]: I think that the original was correct, there are some warnings in py3.8

**Review by Borda — COMMENTED** [review]
we rather move the actual metrics to pl.legacy.metrics then completely before we fill all metrics we had before...
do the same with metric tests aka tests.legacy.metrics

**Comment by ananyahjha93** [comment]
@justusschock we are probably going to deprecate functional metric as with the whole update, reset and compute format, functional metrics don't make sense. If you want a single function to compute your metric value:

```
acc = Accuracy()

for batch in dataloader:
    step_acc_val = acc(preds, target)
```

**Comment by justusschock** [comment]
@ananyahjha93 the functional API does none of these things. It's just batch in, Value out and I really would keep these (same reason why python has finctionals of their loss functions).

File: pytorch_lightning/metrics/metric.py [hunk 8]
```diff
@@ -1,262 +1,168 @@
...
+        elif dist_reduce_fx == "mean":
+            dist_reduce_fx = lambda x: torch.mean(x, dim=0)
+        elif dist_reduce_fx == "cat":
+            dist_reduce_fx = lambda x: torch.cat(x, dim=0)
+        elif dist_reduce_fx is not None and not isinstance(dist_reduce_fx, Callable):
+            raise ValueError(
+                "`dist_reduce_fx` must be callable or one of ['mean', 'sum', 'cat', None]"
+            )
+
+        setattr(self, name, default)
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:69]: What about using `self.register_buffer` here instead of `setattr`?

File: pytorch_lightning/metrics/regression/mean_error.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
+import torch
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/regression/mean_error.py:1]: As stated earlier, I think it would be beneficially to keep each metric in a separate file

**Review by SkafteNicki — COMMENTED** [review]
PR is looking good. Really like many of the changes.

Question: is the idea to do to the hole update in this PR or just lay out the foundation for the metric package here, and the follow up with PRs for individual metrics?

As @justusschock has already mentioned in his review, I see 3 major breaking changes that we at least need to discuss:
1. Do we completely remove the functional interface?
2. Do we remove support for numpy arrays (by removing `NumpyMetric` subclass)?
3. If 2 is true, I assume we also remove sklearn interface?

File: pytorch_lightning/metrics/regression/mean_error.py (hunk 9 above)
Comment (teddykoker) [reply pytorch_lightning/metrics/regression/mean_error.py]: Done!

File: pytorch_lightning/metrics/metric.py [hunk 10]
```diff
@@ -1,262 +1,168 @@
...
-    convert_to_tensor,
-    convert_to_numpy,
-)
 from pytorch_lightning.utilities.apply_func import apply_to_collection
-from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
+from pytorch_lightning.metrics.utils import _flatten, gather_all_tensors_if_available
 
 
-class Metric(DeviceDtypeModuleMixin, nn.Module, ABC):
+class Metric(nn.Module, ABC):
```
[the remark this answers is not in the document]
Comment (teddykoker) [reply pytorch_lightning/metrics/metric.py:16]: Since metrics are `nn.Module`, lightning should be able to move them across devices without issue.

[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:16]: Agree, I was just wondering the reason why we needed it in the first place.

File: pytorch_lightning/metrics/metric.py (hunk 8 above)
Comment (teddykoker) [reply pytorch_lightning/metrics/metric.py]: Alright I added register buffer if the state is a tensor, else it just uses setattr, this will allow the correct device sync behavior

File: pytorch_lightning/metrics/metric.py [hunk 11]
```diff
@@ -1,262 +1,168 @@
...
+                "state variable must be a tensor or any empty list (where you can append tensors)"
+            )
+
+        if dist_reduce_fx == "sum":
+            dist_reduce_fx = lambda x: torch.sum(x, dim=0)
+        elif dist_reduce_fx == "mean":
+            dist_reduce_fx = lambda x: torch.mean(x, dim=0)
+        elif dist_reduce_fx == "cat":
+            dist_reduce_fx = lambda x: torch.cat(x, dim=0)
+        elif dist_reduce_fx is not None and not isinstance(dist_reduce_fx, Callable):
```
[the remark this answers is not in the document]
Comment (ananyahjha93) [reply pytorch_lightning/metrics/metric.py:109]: we are already stacking tensors for each process.

File: pytorch_lightning/metrics/classification/__init__.py [hunk 12]
```diff
@@ -0,0 +1 @@
+from pytorch_lightning.metrics.classification.accuracy import Accuracy
```
[the remark this answers is not in the document]
Comment (ananyahjha93) [reply pytorch_lightning/metrics/classification/__init__.py:1]: doesn't make any sense to export base metric in classification/__init__.py

File: pytorch_lightning/metrics/functional/__init__.py [hunk 13]
```diff
@@ -1,34 +0,0 @@
-from pytorch_lightning.metrics.functional.classification import (
```
[the remark this answers is not in the document]
Comment (ananyahjha93) [reply pytorch_lightning/metrics/functional/__init__.py:1]: not keeping anything functional from older API

File: pytorch_lightning/metrics/functional/classification.py [hunk 14]
```diff
@@ -1,1056 +0,0 @@
-from functools import wraps
```
[the remark this answers is not in the document]
Comment (ananyahjha93) [reply pytorch_lightning/metrics/functional/classification.py:1]: not keeping anything functional from older API

File: pytorch_lightning/metrics/functional/reduction.py [hunk 15]
```diff
@@ -1,65 +0,0 @@
-import torch
```
[the remark this answers is not in the document]
Comment (ananyahjha93) [reply pytorch_lightning/metrics/functional/reduction.py:1]: not keeping anything functional from older API

File: pytorch_lightning/metrics/functional/self_supervised.py [hunk 16]
```diff
@@ -1,46 +0,0 @@
-import torch
```
[the remark this answers is not in the document]
Comment (ananyahjha93) [reply pytorch_lightning/metrics/functional/self_supervised.py:1]: not keeping anything functional from older API

File: pytorch_lightning/metrics/metric.py [hunk 17]
```diff
@@ -1,262 +1,148 @@
...
-        aggr = self.aggregate(*self._step_vals if len(self._step_vals) > 1 else self._step_vals)
-        self.reset()
-        return self.compute(self, None, aggr)
+    def compute(self):  # pylint: disable=E0202
+        pass
 
     def reset(self):
-        self._step_vals = []
+        for attr, default in self._defaults.items():
+            setattr(self, attr, deepcopy(default))
```
[the remark this answers is not in the document]
Comment (ananyahjha93) [reply pytorch_lightning/metrics/metric.py]: caching in self._computed

**Comment by AtlantixJJ** [comment]
Hi guys, I just install the pytorch-lightning package with the latest version in master, but found out that the metrics.functional was gone and the metrics were almost empty.

Is that true for the moment due to the revamping?

Would you recommend a usable version as of now? Thanks
